rolling-update-kubernetes/src/test/scala/akka/rollingupdate/kubernetes/AppVersionRevisionSpec.scala

```diff
@@ -0,0 +1,187 @@
+
+  private def replicaPath(replica: String) =
+    urlEqualTo(s"/apis/apps/v1/namespaces/$namespace/replicasets/$replica")
+
+  private def getPod(podName: String): MappingBuilder =
+    get(podPath(podName)).withHeader("Content-Type", new EqualToPattern("application/json"))
+
+  private def getReplicaSet(replica: String): MappingBuilder =
+    get(replicaPath(replica)).withHeader("Content-Type", new EqualToPattern("application/json"))
+
+  private val defaultPodResponseJson =
+    """{
+      | "metadata": {
+      |   "ownerReferences": [
+      |     {"name": "wrong-replicaset-id", "kind": "SomethingElse"},
+      |     {"name": "parent-replicaset-id", "kind": "ReplicaSet"}
+      |    ]
+      |  }
+      |}""".stripMargin
+
+  private val defaultReplicaResponseJson =
+    """{
+      | "metadata": {
+      |   "annotations": {
+      |     "deployment.kubernetes.io/revision": "1"
+      |   }
+      |  }
+      |}""".stripMargin
+
+  private def stubPodResponse(json: String = defaultPodResponseJson) =
+    stubFor(
+      getPod(podName1).willReturn(
+        ResponseDefinitionBuilder.okForJson("").withJsonBody(mapper.readTree(json))
+      ))
+
+  private def stubReplicaResponse(json: String = defaultReplicaResponseJson) =
+    stubFor(
+      getReplicaSet("parent-replicaset-id").willReturn(
+        ResponseDefinitionBuilder.okForJson("").withJsonBody(mapper.readTree(json))
+      ))
+
+  "Read revision from Kubernetes" should {
+
+    "parse pod and replica response to get the revision" in {
+      stubPodResponse()
+      stubReplicaResponse()
+
+      EventFilter
+        .info(pattern = "Reading revision from Kubernetes: akka.cluster.app-version was set to 1", occurrences = 1)
+        .intercept {
+          kubernetesApi.readRevision()
+        }
+    }
+
+    "retry and then fail when pod not found" in {
+      stubFor(getPod(podName1).willReturn(aResponse().withStatus(404)))
+      EventFilter
+        .warning(pattern = ".*Failed to get revision", occurrences = 5)
+        .intercept({
+          kubernetesApi.readRevision()
+        })
+    }
+
+    "retry and then fail when replicaset not found" in {
+      stubPodResponse()
+      stubFor(getReplicaSet("parent-replicaset-id").willReturn(aResponse().withStatus(404)))
+      EventFilter
+        .warning(pattern = ".*Failed to get revision", occurrences = 5)
+        .intercept({
+          kubernetesApi.readRevision()
+        })
+    }
+
+    "log if pod json can not be parsed" in {
+      stubPodResponse(json = """{ "invalid": "json" }""")
+      EventFilter
+        .warning(pattern = ".*Error while parsing Pod*")
+        .intercept({
+          kubernetesApi.readRevision()
+        })
+    }
+
+    "log if replica json can not be parsed" in {
+      stubReplicaResponse(json = """{ "invalid": "json" }""")
+      EventFilter
+        .warning(pattern = ".*Error while parsing Pod*")
+        .intercept({
+          kubernetesApi.readRevision()
+        })
+    }
+  }
+}
```

Comment:
No test of retry on failure and then succeeding on retry? Tricky to express with wireMock?